Add support for Shadow DOM (ShadowRoot)

zm-cttae opened the pull request:
Resolves #86.

zm-cttae committed:
Add open shadow DOM cloning support (#86)

meche-gh commented:
Possible test case we can comandeer: 

<details>
<summary><code>./test-lib/mdn-web-components-examples/slotchange.html</code> <a href="https://mdn.github.io/web-components-examples/slotchange/">(source)</a></summary>

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>mdn/web-components-examples/slotchange</title>
    <script defer>
      customElements.define('summary-display',
        class extends HTMLElement {
          constructor() {
            super();
            const template = document.getElementById('summary-display-template');
            const templateContent = template.content;
            const shadowRoot = this.attachShadow({mode: 'open'});
            shadowRoot.appendChild(templateContent.cloneNode(true));
            const items = Array.from(this.querySelectorAll('li'));
            const descriptions = Array.from(this.querySelectorAll('p'));
            items.forEach(item => { handleClick(item) });

            const slots = this.shadowRoot.querySelectorAll('slot');
            slots[1].addEventListener('slotchange', function(e) {
              const nodes = slots[1].assignedNodes();
              console.log(`Element in Slot "${slots[1].name}" changed to "${nodes[0].outerHTML}".`);
            });

            function handleClick(item) {
              item.addEventListener('click', function() {
                items.forEach(item => {
                  item.style.backgroundColor = 'white';
                });
                descriptions.forEach(description => {
                  updateDisplay(description, item);
                });
              });
            }

            function updateDisplay(description, item) {
              description.removeAttribute('slot');

              if(description.getAttribute('data-name') === item.textContent) {
                description.setAttribute('slot', 'choice');
                item.style.backgroundColor = '#bad0e4';
              }
            }
          }
        }
      );
    </script>
    <style>
      li {
        cursor: pointer;
      }
    </style>
  </head>
  <body>
    <h1>slotchange event example</h1>

    <summary-display>
      <ul slot="master-list">
        <li>Apples</li>
        <li>Pears</li>
        <li>Bananas</li>
        <li>Oranges</li>
        <li>Peaches</li>
        <li>Strawberries</li>
        <li>Blueberries</li>
      </ul>

      <p data-name="Apples">A common, sweet, crunchy fruit, usually green or yellow in color.</p>
      <p data-name="Pears">A fairly common, sweet, usually green fruit, usually softer than Apples.</p>
      <p data-name="Bananas">A long, curved, yellow fruit, with a fairly gentle flavor.</p>
      <p data-name="Oranges">Orange in color, usually sweet but can be sharp, often contains pips.</p>
      <p data-name="Peaches">An orange fruit with big stone in the middle, and sweet, juicy flesh.</p>
      <p data-name="Strawberries">A red fruit with yellow seeds on the outside; has a sweet flavor and a pretty shape.</p>
      <p data-name="Blueberries">They are berries and they are blue; sweet in flavor, small in size, round.</p>
    </summary-display>

    <template id="summary-display-template">
      <article>
        <div>
          <slot name="master-list"></slot>
        </div>
        <div>
          <slot name="choice"></slot>
        </div>
      </article>
    </template>
  </body>
</html>
```

</details>

zm-cttae committed:
Revert filtering out shadow DOM `<script>`s

IDisposable commented on src/dom-to-image-more.js:
```diff
@@ -236,27 +235,41 @@
 
     function cloneNode(node, filter, root, parentComputedStyles, ownerWindow) {
-        if (!root && filter && !filter(node)) {
+        if (!root && (filter && !filter(node))) {
```
I might have fixed part of this over in #104 by making `cloneNode` skip any node with a tagName of  `'SCRIPT'`...

meche-gh replied:
Best to merge #104 to master and then deal with the line in the merge conflict.

meche-gh replied:
No longer in the PR diff! :)

IDisposable commented on src/dom-to-image-more.js:
```diff
@@ -954,7 +987,7 @@
             sandbox.contentDocument.title = 'sandbox';
         }
         const defaultElement = document.createElement(tagName);
```
In #104, I changed this line to (essentially) `sandbox.contentDocument.createElement(tagName);` because I think we need to be creating the element against the sandbox document NOT against the `document`...

meche-gh replied:
Yep. Another one for the merge conflict :)

zm-cttae committed:
Merge branch 'master' into feat-shadow-dom-support

IDisposable commented:
Hate to do this to you, but probably will land this after the #106... do you have a test we can put in the suite?

zm-cttae committed:
Filter out `<script>` elements to keep DOM static
Refs: 1904labs/dom-to-image-more#104

zm-cttae committed:
Add tests for #86

meche-gh commented:
Added tests, greatly simplified that snippet first too :)

zm-cttae committed:
Add timeout to #86 test

IDisposable committed:
Merge branch 'master' into feat-shadow-dom-support

IDisposable commented on src/dom-to-image-more.js:
```diff
@@ -257,24 +258,38 @@
+        }
+
+        function cloneShadowNode(original) {
+            const nextSibling = original.nextSibling;
```
Will `original.nextSibling` being null/nothing ever be an issue down on line 283?

meche-gh replied:
Don't worry about it - [Node.insertBefore() - Web APIs | MDN (mozilla.org)](https://developer.mozilla.org/en-US/docs/Web/API/Node/insertBefore#example_2)

meche-gh replied:
Line removed in d656297 regardless

IDisposable commented on src/dom-to-image-more.js:
```diff
@@ -257,24 +258,38 @@
+            const nextSibling = original.nextSibling;
+            const parentNode = original.parentNode;
+            // Bypass element encapsulation, by detaching from shadow node then restoring.
+            parentNode.removeChild(original);
```
This seems like we're changing the original DOM, which is something we've never intentionally done before. Would this be a problem? Are we changing anything that the code/DOM on the clonee side would care about?

meche-gh replied:
*   We can't clone from the original shadow DOM unfortunately (`cloneNode` errors out).
*   This isn't that impactful to the end user - the change is reversed and the `Node` is restored.
*   We will be triggering MutationObserver callbacks by doing this.  
    I believe this might be an issue with SPAs that render views for a model collection in an app.  
    (Thinking lists or outlines e.g. chat users in a channel list, code symbols in VS Code's "Outline".)  
    However, if those apps are not taking ownership of those callbacks and shelling out to MutationObserver, that's arguably a poor design decision on their part.

meche-gh replied:
Turns out the logic to detach from the shadow DOM is only ever needed if you are trying to run `cloneNode` method on a `#shadow-root` fragment. I'll remove the logic sharpish.

meche-gh replied:
Resolved in d656297

IDisposable commented on src/dom-to-image-more.js:
```diff
@@ -257,24 +258,38 @@
+            // Bypass element encapsulation, by detaching from shadow node then restoring.
+            parentNode.removeChild(original);
+            const clone = original.cloneNode(false);
+            parentNode.insertBefore(original, nextSibling);
```
Does this work if there was no `nextSibling`?

meche-gh replied:
Yup - https://developer.mozilla.org/en-US/docs/Web/API/Node/insertBefore#example_2

meche-gh replied:
Lines removed in d656297 regardless

IDisposable commented on src/dom-to-image-more.js:
```diff
@@ -257,24 +258,38 @@
 
             if (originalChildren.length !== 0) {
-                const originalComputedStyles = getComputedStyle(original);
+                const originalComputedStyles = getComputedStyle(util.isShadowRoot(original) ? original.host : original);
```
Is it simply enough to do `original.host || original`? I'm really not familiar with the shadow DOM semantics as a whole, but this seems like a lot of fragile checks. I would welcome a chat about these changes.

meche-gh replied:
If `original` is a shadow DOM root (due to the ternary at line 261), we need to pass through the parent styling of its root element.

Let's take `<relative-time>` as the example.
In line 261 we "jump" from `<relative-time>` to its `#shadow-root` so we can copy its children (the text "X days ago"). The `<relative-time>` element can be accessed from `child=TextNode` by `child.parentNode.host`.

For maintainability I will put in extra variables (ternary output as a function call param is not that great). It's probably going to need explanation anyway so a one-liner comment would be good.

meche-gh replied:
Resolved in d656297

IDisposable commented on src/dom-to-image-more.js:
```diff
@@ -257,24 +258,38 @@
 
         function cloneChildren(original, clone) {
-            const originalChildren = original.childNodes;
+            const originalChildren = util.isShadowSlotElement(original) ? original.assignedNodes() : original.childNodes;
```
Is `original.assignedNodes()` null or just zero-lengthed if this is NOT a shadow slot element?

meche-gh replied:
It's null :s

meche-gh replied:
It's null :s

meche-gh replied:
Abstracted away and added a comment in [d656297](https://github.com/1904labs/dom-to-image-more/pull/101/commits/d65629752cbacad624ac790d3bec9980b553361c)

IDisposable commented on src/dom-to-image-more.js:
```diff
@@ -298,7 +313,8 @@
 
         function processClone(original, clone) {
-            if (!util.isElement(clone)) {
+            if (!util.isElement(clone)) { return Promise.resolve(clone); }
```
This may fix something I randomly saw broken, hmmm

meche-gh replied:
The `Promise.resolve` was there before - I didn't see a rationale to kill it so took the cautious route and changed it back

IDisposable commented on src/dom-to-image-more.js:
```diff
@@ -298,7 +313,8 @@
         function processClone(original, clone) {
-            if (!util.isElement(clone)) {
+            if (!util.isElement(clone)) { return Promise.resolve(clone); }
+            if (!util.isElement(clone) || util.isShadowSlotElement(original)) {
```
Not sure how _both_ of these lines are supposed to be here... might be a merge issue.

meche-gh replied:
Resolved in 3c6b9c7

zm-cttae committed:
Finish `processClone` merge resolve for early conditional return

meche-gh commented on src/dom-to-image-more.js:
```diff
@@ -249,6 +249,7 @@
         const filter = options.filter;
         if (
             node === sandbox ||
+            util.isHTMLScriptElement(node) ||
```
We made a conscious decison to kill `SCRIPT` elements to better freeze the DOM, but we would also benefit from killing `STYLE` and `LINK`. ([The people who worship Shadow DOM](https://github.com//lit/lit-element/issues/553) like to flood the shadow element with these tags and call it "encapsulation". </scathe>)

- The `STYLE` tags duplicate existing styling data and also add `@media` queries that shouldn't be there (because they change the cloned output in bad ways when the browser window is resized). These media queries really get in the way of debugging SVG output files.
- We can also remove `LINK` elements seeing as they are broken by the cloning process.

Thoughts @IDisposable?

IDisposable replied:
Yes, we certainly ought to not clone `STYLE` tags as they pollute our emitted SVG to no effect. Never seen a `LINK` in the wild with this either, so I'm good with skipping those as well.

zm-cttae committed:
Refactor shadow DOM logic
- Remove extraneous logic to handle ShadowRoot
- Abstract `cloneChildren` ternaries awayy

zm-cttae committed:
Clean up pass for #101

IDisposable commented on src/dom-to-image-more.js:
```diff
@@ -257,24 +258,33 @@
+            return original.cloneNode(false);
+        }
+
+
```
_nit_ extra blank line

zm-cttae committed:
Filter out `STYLE` and `LINK` tags for #101
These tags are (mis?)used without scoping in shadow DOM
Remove them to prevent the cloned shadow DOM breaking
Also STYLE tags can contribute media queries that damage output integrity

zm-cttae committed:
Cleanup extra newline left in `cloneNode`

IDisposable commented on src/dom-to-image-more.js:
```diff
@@ -512,6 +541,18 @@
         }
 
+        function hasOpenShadowRoot(value) {
+            return isElement(value) && value.shadowRoot !== null;
```
Weird that on this method we're hitting the property `.shadowRoot` of the element directly, while in line 549, we're calling `getWindow()` on the element and asking if the `ShadowRoot` instanceof matches value... I guess I'm wondering if an element's `shadowRoot` _is_ itself when it's a shadow root... e.g. can we do `return value === value.shadowRoot;`? I'm really weak on this whole area, sorry.

IDisposable replied:
Or... looking at line 553, could/should _this_ line be `return isInShadowRoot(value) && value === value.getRootNode();`

meche-gh replied:
The name of the function should probably be `isElementHostForOpenShadowRoot`

IDisposable commented:
This is looking REALLY good, thanks for all your hard work :)
The questions I have outstanding are trivial and you can just say :shipit: and I'm happy to merge as we are now.

zm-cttae committed:
Rename shadowRoot property check for #101

meche-gh commented:
# :shipit:

IDisposable commented:
This seems to have failed unit testing
<img width="615" alt="image" src="https://user-images.githubusercontent.com/874059/214696629-0b94a99b-5e94-43e7-bcfe-5ebf9d866792.png">

meche-gh commented:
Yup must be that margin on `<p>`

IDisposable commented:
The whole cloned node section is blank.... creating a fiddle
https://jsfiddle.net/w0efg2y7/6/

IDisposable commented:
This is the emitted IFRAME from the fiddle
[shadow-fiddle.html.txt](https://github.com/1904labs/dom-to-image-more/files/10504225/shadow-fiddle.html.txt)
(rename back to HTML)

meche-gh commented:
I think there must have been a race condition where the `<script>` kicked in after the test bench ran. It's a custom element that won't render until then

Maybe it works if the <script> tag is inside the element

IDisposable commented:
I moved the script up to above the content and now we get a warning when running it... but the cloning works. SO WRONG
https://jsfiddle.net/c349vyLs/4/

meche-gh commented:
Got it ... the element in the testcase is technically dependent on `template` also so it needs to go after it.

IDisposable commented:
... going to have to think through how to delay long enough to make the tests pass... maybe a simple await  somewhere.
